packages/plugin-constraints/sources/tauModule.ts

```diff
@@ -0,0 +1,68 @@
+import {Project} from '@berry/core';
+import pl        from 'tau-prolog';
+
+const {is_atom} = pl.type;
+
+function prependGoals(
+    thread: pl.type.Thread, point: pl.type.State,
+    goals: pl.type.Term<number, string>[]): void {
+  thread.prepend(goals.map(
+      goal => new pl.type.State(
+          point.goal.replace(goal), point.substitution, point)));
+}
+
+const projects = new WeakMap<pl.type.Session, Project>();
+
+function getProject(thread: pl.type.Thread): Project {
+  const project = projects.get(thread.session);
+
+  if (project == null) {
+    throw new Error(`No project registered for the active session`);
+  }
+
+  return project;
+}
+
+const tauModule = new pl.type.Module(
+    `constraints`, {
+      'workspace_field/3': (thread, point, atom) => {
+        const [workspaceCwd, fieldName, fieldValue] = atom.args;
+
+        if (!is_atom(workspaceCwd) || !is_atom(fieldName)) {
+          thread.throw_error(pl.error.instantiation(atom.indicator));
+          return;
+        }
+
+        const workspace =
+            getProject(thread).tryWorkspaceByCwd(workspaceCwd.id);
+
+        if (workspace == null) {
+          // Workspace not found => this predicate can never match
+          // We might want to throw here? We can be pretty sure the user did
+          // something wrong at this point
+          return;
+        }
+
+        const manifest: {[key: string]: any} = {};
+        workspace.manifest.exportTo(manifest);
```

Comment:
Rather than use `exportTo` (which requires extra copies), what do you think of instead using `Object.prototype.hasOwnProperty.call` on the manifest instance itself?

Reply:
I was thinking about this as well&hellip; Would it be possible to store the "raw" manifest in the `Manifest` type?
The `Manifest` class only exposes supported properties. If someone wants to add a custom property specifically for the constraints check, that wouldn't work. For example, in a library in two parts, `core` and `ui`, and `core` workspaces musn't depend on `ui` workspaces, I might want to add a custom field to my manifest and use that in the constraints file rather than do complicated string checks on package idents or cwds.

Reply:
That `raw` field seems like a pretty good idea! 👍